server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/CheckRunningJob.java

```diff
@@ -54,51 +62,86 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
-            doHandleJobRetryingOrFailed(a);
-        });
-
+        for (JobEntity j : jobs) {
+            SilentExecutor.executeSafely("handleJobRetryingOrFailed", () -> {
+                handleJobRetryingOrFailed(j, a -> {
+                    try {
+                        // destroy executor
+                        getConfiguration().getJobDispatcher().destroy(JobIdentity.of(a.getId()));
+                    } catch (JobException e) {
+                        throw new TaskRuntimeException(e);
+                    }
+                });
+            });
+        }
     }
 
-    private void doHandleJobRetryingOrFailed(JobEntity a) {
-        // destroy executor
-        try {
-            getConfiguration().getJobDispatcher().destroy(JobIdentity.of(a.getId()));
-        } catch (JobException e) {
-            throw new TaskRuntimeException(e);
+    private void handleNotLocalAndProcessIsNotExitsJobs(int size, int heartTimeoutPeriod) {
+        // ip has been changed docker restart and process has been interrupted,
+        // so we should find jobs which heart timeout and not running in local
+        Page<JobEntity> jobs = getConfiguration().getTaskFrameworkService()
+                .findHeartTimeTimeoutNotLocalJobs(heartTimeoutPeriod, 0, size);
+
+        List<JobEntity> jobList = jobs.stream().filter(a -> {
+            ExecutorIdentifier ei = ExecutorIdentifierParser.parser(a.getExecutorIdentifier());
+            return !(HttpUtil.isConnectable(ei.getHost(), ei.getPort(),
+                    getConfiguration().getTaskFrameworkProperties().getCheckOdcServerCanBeConnectedTimes()));
+        }).collect(Collectors.toList());
+
+        for (JobEntity j : jobList) {
+            SilentExecutor.executeSafely("handleJobRetryingOrFailed", () -> {
+                handleJobRetryingOrFailed(j, a -> {
+                    int rows = getConfiguration().getTaskFrameworkService().updateExecutorToDestroyed(a.getId());
+                    if (rows > 0) {
+                        log.info("Executor is not exists, update job executor to destroyed, jobId={}", a.getId());
+                    } else {
+                        throw new TaskRuntimeException("update executor to destroyed failed, jobId=" + a.getId());
+                    }
+
+                });
+            });
         }
+    }
 
-        if (checkJobIfRetryNecessary(a)) {
-            log.info("Need to restart job, destroy old executor completed, jobId={}.", a.getId());
-            int rows = getConfiguration().getTaskFrameworkService()
-                    .updateStatusDescriptionByIdOldStatusAndExecutorDestroyed(a.getId(), JobStatus.RUNNING,
-                            JobStatus.RETRYING, "Heart timeout and retrying job");
-            if (rows > 0) {
-                log.info("Job {} set status to RETRYING.", a.getId());
+    private void handleJobRetryingOrFailed(JobEntity a, Consumer<JobIdentity> jobIdentityConsumer) {
```

Comment:
why should we need this consumer? does this consumer has another implementation ?

Reply:
yes , `handleGeneralHeartTimeoutJobs` and `handleNotLocalAndProcessIsNotExitsJobs` has two implementation